File: 感兴趣的算法/同态加密.py
def isprime_mr(a,b=None):
    if b is None: b = [2,3,5,7,11,13,17,19,23,29,31,37,41,43,47,53,59,61,67,71,73,79,83,89,97]
    if a == 2:return True
    if a%2==0 or a==1:return False
    for x in b:
        t = a - 1
        while t%2 == 0:
            v = pow(x,t,a)# python自带函数pow就带有快速幂取模功能
            if v not in [0,1,a-1]:
                return False
            else:
                if v in [0,a-1]: break
                t //= 2
    return True
def make_primes(length=1024, e=65537):
    if e == 2: raise KeyError('The parameter E must not be equal to 2.')
    import random
    def get_prime(bitlen=1024):
        num = (1<<(bitlen-1)) + random.randint(0,1<<(bitlen-1)) | 1
        while True:
            num += 2
            if isprime_mr(num):
                return num
    while True:
        p = get_prime(length//2)
        q = get_prime(length//2)
        n = p * q
        if n.bit_length() == length and p != q:
            break
    return n,p,q


bits = 64
n,p,q = make_primes(bits)
n2 = n ** 2
print('n,p,q,n2:', n,p,q,n2)

def ex_gcd(a,b):
    if b == 0:
        return (1,0,a)
    (x, y, r) = ex_gcd(b,a%b)
    t = x
    x = y
    y = t - a//b*y
    return (x,y,r)

def modinv(a, m):
    x, y, g = ex_gcd(a, m)
    if g != 1:
        raise Exception('modular inverse does not exist')
    else:
        return x % m

def lcm(a, b):
    return a * b // ex_gcd(a, b)[-1]




import random

lamb = lcm(p-1, q-1)
x = modinv((pow(n + 1, lamb, n2) - 1) // n, n)
print(lamb, x)

def encrypt(m):
    return (n * m + 1) % n2

def add(a, b):
    return (a * b) % n2

def mult(a, b):
    return pow(a, b, n2)

def decrypt(c):
    return ((pow(c, lamb, n2) - 1) // n * x) % n

e1 = encrypt(2)
e2 = encrypt(3)
print(e1, e2)

e3 = add(e1, e2)
c3 = decrypt(e3)

e4 = mult(e1, 3)
c4 = decrypt(e4)

print('n2:', n2)
print('e3:', e3)
print('c3:', c3)
print('e4:', e4)
print('c4:', c4)

File: 感兴趣的算法/同态加密.js
function quickpow(m, n, k){
    var r = 1;
    while(n > 0) {
        if(n & 1)
            r = (r * m) % k;
        n = n >> 1 ;
        m = (m * m) % k;
    }
    return r;
}

// miller-rabin 算法素性检测
function isprime_mr(a,b){
    if (!b){
        b = [2,3,5,7,11,13,17,19,23,29,31,37,41,43,47,53,59,61,67,71,73,79,83,89,97]
    }
    if (a == 2){
        return true
    }
    if (a % 2 == 0 || a == 1){
        return false
    }
    for (var i = 0; i < b.length; i++) {
        var x = b[i]
        var t = a - 1
        while (t % 2 == 0){
            var v = quickpow(x, t, a)
            if (!(v == 0 ||
                  v == 1 ||
                  v == a-1 )){
                return false
            }else{
                if (v == 0 ||
                    v == a-1){
                    break
                }
                t = (t / 2) | 0
            }
        }
    }
    return true
}

function gen_prime(bitlen){
    var n = (1<<(bitlen-1)) + Math.random() * ((1<<(bitlen-1)) * 0.9) | 1
    while (1){
        n += 2
        if (isprime_mr(n)){
            return n
        }
    }
}

function ex_gcd(a, b){
    if (b == 0){
        return [1, 0, a]
    }
    var z = ex_gcd(b, a%b)
    var x = z[0]
    var y = z[1]
    var r = z[2]
    var t = x
    var x = y
    var y = t - (a / b | 0) * y
    return [x, y, r]
}

function make_primes(l){
    var e = 65537
    if (l % 2 == 1){ throw Error('must be even.') }
    while(1){
        var p = gen_prime(l / 2)
        var q = gen_prime(l / 2)
        var n = p * q
        if ((n.toString(2).length == l) && p != q){
            break
        }
    }
    return [n, p, q]
}

function modinv(a, m) {
    var z = ex_gcd(a, m)
    var x = z[0]
    var y = z[1]
    var g = z[2]
    if (g != 1){
        throw Error('inver not exist.')
    }else{
        if (x > 0){
            return x % m
        }else{
            while (x < 0){
                x += m
            }
            return x
        }
    }
}

function lcm(a, b){
    return a * b / ex_gcd(a, b)[2]
}


var z = make_primes(12)
var n = z[0]
var p = z[1]
var q = z[2]
var n2 = n * n
var lamb = lcm(p-1, q-1)
var x = modinv((quickpow(n + 1, lamb, n2) - 1) / n, n)
console.log(n2, lamb, x)




function encrypt(m){
    return (n * m + 1) % n2
}

function add(a, b){
    return (a * b) % n2
}

function mult(a, b){
    return quickpow(a, b, n2)
}

function decrypt(c){
    return ((quickpow(c, lamb, n2) - 1) / n * x) % n
}




// 适用于运算结果在 1000 以内的整数算法 (+*)
e1 = encrypt(1)
e2 = encrypt(2)
e3 = encrypt(3)
console.log(e1, e2, e3)

// 1 + 2 + 3
e4 = add(add(e1, e2), e3)
c4 = decrypt(e4)

// (1+2+3) * 5
e5 = mult(e4, 5)
c5 = decrypt(e5)

console.log('n2:', n2)
console.log('e4:', e4)
console.log('c4:', c4)
console.log('e5:', e5)
console.log('c5:', c5)